storage/innobase/row/row0mysql.cc

```diff
@@ -2174,6 +2192,90 @@ row_mysql_unfreeze_data_dictionary(
 	trx->dict_operation_lock_mode = 0;
 }
 
+/** Function restores btr_pcur_t, creates dtuple_t from rec_t,
+sets row_end = CURRENT_TIMESTAMP/trx->id, inserts it to a table and updates
+table statistics.
+This is used in UPDATE CASCADE/SET NULL of a system versioning table.
+@param[in]	thr	current query thread
+@param[in]	node	a node which just updated a row in a foreign table
+@return DB_SUCCESS or some error */
+UNIV_INLINE
+dberr_t
+row_update_versioned_insert(que_thr_t* thr, const upd_node_t* node)
+{
+	const trx_t* trx = thr_get_trx(thr);
+	dict_table_t* table = node->table;
+	ut_ad(table->versioned());
+	dict_index_t* index = dict_table_get_first_index(table);
+
+	mtr_t mtr;
+	mtr.start();
+	mtr.set_named_space(index->space);
+	ut_a(node->pcur->rel_pos == BTR_PCUR_ON);
+
+	ulint mode = 0;
+	if (dict_index_is_online_ddl(index)) {
+		ut_ad(table->id != DICT_INDEXES_ID);
+		mode = BTR_MODIFY_LEAF | BTR_ALREADY_S_LATCHED;
+		mtr_s_lock(dict_index_get_lock(index), &mtr);
+	} else {
+		mode = BTR_MODIFY_LEAF;
+	}
+
+	bool success = btr_pcur_restore_position(mode, node->pcur, &mtr);
+	mtr.commit();
+
+	if (!success) {
+		return DB_RECORD_NOT_FOUND;
+	}
```

Comment:
Why do you think it is OK to `mtr.commit()` after restoring the cursor position?
What is the purpose of restoring the cursor position if you are committing the mini-transaction right after?
Between the `mtr.commit()` and the subsequent mini-transactions that would modify the tree, the state can be changed by other threads. So, the value of `success` is stale.

This check should be done somewhere inside `row_ins_step()` instead. Why not just handle a duplicate key error?

Reply:
What I'm trying to do at this point is to get an old values for updated row to compose historical row. This is the simplest way I found to do it, thanks to `btr_pcur_t` which points to the right place.

Another way to get and old values I considered was reading undo log right update row was updated.

Reply:
I guess I understand why record lock is needed here: it's a shared data after all. I need to lock it and `mtr.commit()` after `row_build()`.

But I do not understand how can I move `btr_pcur_restore_position()` inside `row_ins_step()`: I need to compose a row to insert before this call, not inside of it.

I also do not understand how to get and how to use a duplicate key error here.